XEP-0333: Chat Markers

dommaes opened the pull request:
issue: https://github.com/jsxc/jsxc/issues/745
XEP: https://xmpp.org/extensions/xep-0333.html

sualko commented:
> My idea is to implement XEP-0333 as a plugin

Therefore it would be great if you put your plugin inside the `src/plugins/` folder.

dommaes commented:
I think I'm finally done with this XEP/pull request.
Would you please consider accepting it?
Thanks a lot!

sualko commented on src/plugins/chatMarkers/ChatMarkersPlugin.ts:
```diff
@@ -0,0 +1,333 @@
+            this.addMarkable(xmlStanza, msg);
+         }
+      });
+      return Promise.resolve([msg, xmlStanza]);
```
Modifications from addMarkable are lost at this point. Return this.supportsChatMarkers and return the array inside the then function.

sualko commented on src/plugins/chatMarkers/ChatMarkersPlugin.ts:
```diff
@@ -0,0 +1,333 @@
+
+export default class ChatMarkersPlugin extends AbstractPlugin {
+   public static readonly PATH_CLOSED_ENVELOPE_GREEN =
+      '../../../images/plugins/chatMarkers/chatmarkers_closed_envelope_green.svg' as String;
```
You should use the require syntax here.

dommaes commented:
I implemented your requested changes and squashed the commits into one. Unfortunately I'm not able to resolve the conflict, as Github keeps being unable to merge the changes in src/bootstrap/plugins.ts ...

sualko commented:
> as Github keeps being unable to merge the changes

I think something went wrong during the squash, because your pr is changing 192 files, including several deletions.

dommaes commented:
I managed to squash the commits into one.
Again, would you please consider accepting it?
Thank you!

sualko commented on src/plugins/chatMarkers/ChatMarkersPlugin.ts:
```diff
@@ -0,0 +1,410 @@
+
+      this.data.registerHook('lastReceivedMsgId', (lastReceivedMsgId: string) => {
+         let from = new JID(this.data.get('lastReceivedFrom'));
+         this.supportsChatMarkers(from).then(() => {
```
The result of `supportsChatMarkers` is ignored.

sualko commented on src/plugins/chatMarkers/ChatMarkersPlugin.ts:
```diff
@@ -0,0 +1,410 @@
+   }
+
+   // send "received" on "markable" message according to XEP-0333
+   private onMarkableMessage = (stanza: string) => {
```
Section 9 says "It is possible for a sender to leak its presence when updating Chat Markers; therefore, a sender SHOULD NOT send Chat Markers to recipients who are not otherwise authorized to view its presence.". How is this ensured?

sualko replied:
>  How is this ensured?

I still see no solution for this.

sualko commented on src/plugins/chatMarkers/ChatMarkersPlugin.ts:
```diff
@@ -0,0 +1,410 @@
+         this.chatWindowInitializedHook(chatWindow, contact);
+      });
+
+      this.data.registerHook('lastReceivedMsgId', (lastReceivedMsgId: string) => {
```
This hook is executed in every tab, which results in multiple messages. Why are you using hooks anyway?

sualko commented on src/plugins/chatMarkers/ChatMarkersPlugin.ts:
```diff
@@ -0,0 +1,410 @@
+
+   private setMessagesChatMarkersAcknowledged(lastAcknowledgedMsgUid: string) {
+      let contact = this.pluginAPI.getContact(new JID(this.data.get('lastAcknowledgedFrom')));
+      if (contact !== undefined) {
```
This comparison is not needed, because `undefined` is falsy.

sualko commented on src/plugins/chatMarkers/ChatMarkersPlugin.ts:
```diff
@@ -0,0 +1,410 @@
+   }
+
+   // mark messages, depending on the chat marker we get, according to XEP-0333
+   private onChatMarkersMessage = (stanza: string) => {
```
This function is huge and could be improved.

sualko commented on src/plugins/chatMarkers/ChatMarkersPlugin.ts:
```diff
@@ -0,0 +1,410 @@
+      let transcript = chatWindow.getTranscript();
+      let msg = transcript.getFirstMessage();
+      while (!!msg) {
+         let chatWindowMsg = chatWindow.getChatWindowMessage(msg) as ChatWindowMessage;
```
Why is the cast needed?

dommaes replied:
For type safety, as ChatWindow.getChatWindowMessage() doesn't declare a return type. But as this method returns a ChatWindowMessage for sure, it's probably not necessary to cast.

sualko replied:
The better solution would be to change the signature of `ChatWindow.getChatWindowMessage()` :wink:

sualko commented on src/plugins/chatMarkers/ChatMarkersPlugin.ts:
```diff
@@ -0,0 +1,410 @@
+         element.addClass('jsxc-chatmarkers-acknowledged');
+         element.append($(ChatMarkersPlugin.INDICATOR_TEMPLATE({
+            title: Translation.t('chatmarkers-acknowledged'),
+            src: '../../../images/plugins/chatMarkers/chatmarkers_check_mark_green.svg'
```
Please use the `require` function, because the path can change in some setups.

dommaes commented:
Next try. Would you please take a look and consider it?
Thanks!

sualko commented on src/plugins/chatMarkers/ChatMarkersPlugin.ts:
```diff
@@ -0,0 +1,466 @@
+   private static readonly INDICATOR_TEMPLATE = require('../../../template/plugins/chatMarkers/chatmarkers-indicator.hbs');
+   private static readonly ACKNOWLEDGE_TEMPLATE = require('../../../template/plugins/chatMarkers/chatmarkers-acknowledge.hbs');
+
+   protected pluginAPI = this.pluginAPI as PluginAPI;
```
Why is this necessary?

sualko commented on src/plugins/chatMarkers/ChatMarkersPlugin.ts:
```diff
@@ -0,0 +1,466 @@
+   private setMessagesChatMarkersAcknowledged(lastAcknowledgedMsgUid: string, from: IJID) {
+      this.pluginAPI.Log.debug(`setting previous messages ${ACKNOWLEDGED}. Yaay! =)`);
+      let contact = this.pluginAPI.getContact(from);
+      if (!!contact) {
```
It's not required to have two exclamation marks here, but it would reduce the indentation of the following code if you return if there is no contact.

sualko commented on src/plugins/chatMarkers/ChatMarkersPlugin.ts:
```diff
@@ -0,0 +1,466 @@
+      if (msg.getType() === Message.MSGTYPE.CHAT) {
+         this.supportsChatMarkers(msg.getPeer()).then((hasFeature) => {
+            if (hasFeature
+               && (this.getPresence() === Presence.online
```
Why do you have to check the presence?

sualko commented on src/plugins/chatMarkers/ChatMarkersPlugin.ts:
```diff
@@ -0,0 +1,466 @@
+   }
+
+   // send "received" on "markable" message according to XEP-0333
+   private onMarkableMessage = (stanza: string) => {
```
Would be great if you could reduce the indentation.

sualko commented on src/plugins/chatMarkers/ChatMarkersPlugin.ts:
```diff
@@ -0,0 +1,466 @@
+   }
+
+   // mark messages, depending on the chat marker we get, according to XEP-0333
+   private markMessages(element: JQuery<HTMLElement>, stanza: string, status: ChatMarkerStatus) {
```
This function could be refactored, because it's confusing.

sualko commented on src/plugins/chatMarkers/ChatMarkersPlugin.ts:
```diff
@@ -0,0 +1,466 @@
+               if (!!contact) {
+                  let transcript = contact.getChatWindow().getTranscript();
+                  let msg = transcript.getMessage(msgId);
+                  while (!!msg) {
```
Looks similar to `setMessagesChatMarkersAcknowledged`

dommaes commented:
I think this time I got, what section 9 is referencing to. I did implement your requested changes, as well.
Would you please consider the PR again?

sualko commented:
Code looks good, but I can't merge it, because currently I have no time to test this pr and we are already in feature freeze for 4.0. Therefore this feature will go into 4.1. Thanks for your work :+1:.

sualko commented:
I made my changes in https://github.com/jsxc/jsxc/tree/dommaes-xep0333

Can you test that branch? There are still two things missing:

- [ ] get markers from forwarded messages (carbon copy)
- [ ] get markers from mam

sualko commented:
Markers are now correctly synced and we have a new "transferred" state, like in many other chat clients. I manually merged your changes. Thanks for your contribution.

![2020-04-18 11 03 09 localhost 6ed81923c3ec](https://user-images.githubusercontent.com/2974196/79635951-79ed3680-8174-11ea-9e5d-8234c9fcbc4c.png)
displayed, received, transferred (top to bottom)
